File: tools/contract.py
#!/usr/bin/env python3
"""Read-only campus contract resolvers. Stock Python; no campus identity authority.

Use Resolver(path) for an explicit contract or the module-level functions for
CAMPUS_CONTRACT / the contract beside this tool's own tree.
platform: codex; seat: codex
"""
from __future__ import annotations
import datetime as dt
from functools import lru_cache
import json
import os
from pathlib import Path
import re
import sys


class ContractError(Exception):
    pass


def _require(value):
    if not value:
        raise ValueError('shape')


def _invalid_constant(value):
    raise ValueError(f'invalid JSON constant: {value}')


def contract_path(override=None):
    return Path(override or os.environ.get('CAMPUS_CONTRACT') or
                Path(__file__).resolve().parent.parent / 'contracts/campus.json').resolve()


class Resolver:
    def __init__(self, path=None):
        path = contract_path(path)
        try:
            self.data = json.loads(path.read_text(encoding='utf-8'), parse_constant=_invalid_constant)
        except FileNotFoundError as exc:
            raise ContractError(f'CONTRACT_MISSING: {path}') from exc
        except (OSError, ValueError, UnicodeError) as exc:
            raise ContractError(f'CONTRACT_UNREADABLE: {path}') from exc
        c = self.data
        if not isinstance(c, dict) or c.get('contract_version') != '1.0':
            raise ContractError('CONTRACT_VERSION_UNSUPPORTED: expected contract_version 1.0')
        try:
            self.files = c['context_aliases']['files']
            _require(isinstance(self.files, list))
            for row in self.files:
                _require(isinstance(row['concept'], str) and isinstance(row['canonical'], str))
                _require(isinstance(row['founding_writes'], str))
                _require(isinstance(row['aliases'], list) and all(isinstance(a, str) for a in row['aliases']))
            _require(c['context_aliases']['resolution'] == [
                'canonical', 'legacy alias (in listed order)', 'case-insensitive match'])
            self.hosts = c['hosts']
            _require(isinstance(self.hosts['clients'], dict))
            _require(isinstance(self.hosts['entry_file_by_manifest_dir'], dict))
            _require(isinstance(self.hosts['entry_file_read_names'], list))
            for client in self.hosts['clients'].values():
                _require(isinstance(client['manifest_dir'], str) and isinstance(client['entry_file'], str))
            _require(all(isinstance(v, str) for v in self.hosts['entry_file_by_manifest_dir'].values()))
            self.enums = c['registry_enums']['enums']
            for key in ('department', 'status', 'type'):
                _require(isinstance(self.enums[key], list) and all(isinstance(v, str) for v in self.enums[key]))
            self.aliases = c['registry_enums']['legacy_read_aliases']['department']
            _require(isinstance(self.aliases, dict))
            self.seat = c['seat']
            self.seat_rx = re.compile(self.seat['seat_id_pattern'])
            for key in ('primary', 'satellite', 'unknown'):
                _require(isinstance(self.seat['ledger'][key], str))
            _require(isinstance(self.seat['roles'], list) and isinstance(self.seat['lane']['satellite'], str))
            self.patterns = [(name, re.compile(source)) for name, source in c['run_id']['patterns'].items()]
            _require(self.patterns and self.patterns[0][0] == 'canonical')
            self.segment_rx = re.compile(c['run_id']['segment_pattern'])
            self.dirs = c['run_id']['run_dir']
            prefix, suffix = self.dirs['campus'].split('{team}')
            playbook = self.dirs['playbook']
            _require(playbook.startswith(prefix) and playbook.endswith(suffix) and suffix)
            self.playbook_owner = playbook[len(prefix):-len(suffix)]
            _require(self.segment_rx.fullmatch(self.playbook_owner))
            _require(isinstance(self.dirs['standalone'], str))
        except (KeyError, TypeError, ValueError, AttributeError, re.error) as exc:
            raise ContractError('CONTRACT_INVALID: required resolver data or regex invalid') from exc

    def resolve_context(self, name):
        found = None
        matched = None
        # Global passes: a canonical filename wins over another concept's alias.
        for mode in ('canonical', 'alias', 'case-insensitive', 'concept'):
            for row in self.files:
                names = [row['canonical'], *row['aliases']]
                concepts = [row['concept'], *row['concept'].split('/')]
                concepts += [n.rsplit('.', 1)[0] for n in names]
                match = (name == row['canonical'] if mode == 'canonical' else
                         name in row['aliases'] if mode == 'alias' else
                         name.lower() in [n.lower() for n in names] if mode == 'case-insensitive' else
                         name.lower() in [n.strip().lower() for n in concepts])
                if match:
                    found, matched = row, mode
                    break
            if found:
                break
        return {'concept': found['concept'] if found else None,
                'canonical': found['canonical'] if found else None, 'matched_by': matched,
                'founding_writes': found['founding_writes'] if found else None,
                'read_order': [found['canonical'], *found['aliases']] if found else []}

    def resolve_department(self, value):
        canonical = value if value in self.enums['department'] else self.aliases.get(value)
        known = canonical in self.enums['department']
        return {'value': value, 'canonical': canonical if known else None,
                'legacy_alias': known and value != canonical, 'known': known}

    def _normalize(self, value, kind):
        if value in self.enums[kind]:
            return {'value': value, 'canonical': value, 'known': True}
        for canonical in self.enums[kind]:
            if value.lower() == canonical.lower():
                return {'value': value, 'canonical': canonical, 'known': True,
                        'matched_by': 'case-insensitive'}
        return {'value': value, 'canonical': None, 'known': False}

    def normalize_status(self, value):
        return self._normalize(value, 'status')

    def normalize_type(self, value):
        return self._normalize(value, 'type')

    def entry_file(self, value):
        client = self.hosts['clients'].get(value)
        directory = client['manifest_dir'] if client else value
        entry = self.hosts['entry_file_by_manifest_dir'].get(directory)
        return {'manifest_dir': directory if entry is not None else None,
                'entry_file': entry, 'read_names': list(self.hosts['entry_file_read_names'])}

    def ledger_path(self, info):
        valid = (isinstance(info, dict) and isinstance(info.get('seat'), str) and
                 bool(self.seat_rx.fullmatch(info['seat'])) and info.get('role') in self.seat['roles'])
        state = 'valid' if valid else 'missing' if info is None else 'invalid'
        role, seat = (info['role'], info['seat']) if valid else ('satellite', 'unknown')
        ledger = self.seat['ledger'][role if valid else 'unknown'].format(seat=seat)
        lane = self.seat['lane']['satellite'].format(seat=seat) if role == 'satellite' else None
        return {'state': state, 'role': role, 'seat': seat, 'ledger': ledger, 'lane': lane}

    def parse_run_id(self, run_id):
        for name, rx in self.patterns:
            match = rx.fullmatch(run_id)
            if not match:
                continue
            values = match.groupdict()
            date = values['date'].replace('-', '')
            time = values['time'] + ('00' if len(values['time']) == 4 else '')
            try:
                stamp = dt.datetime.strptime(date + time + (values.get('millis') or '000'),
                                             '%Y%m%d%H%M%S%f').replace(tzinfo=dt.timezone.utc)
            except ValueError:
                break
            canonical = name == 'canonical'
            return {'slug': values.get('slug') or run_id,
                    'stamp_utc': stamp.isoformat(timespec='milliseconds').replace('+00:00', 'Z'),
                    'legacy': not canonical, 'canonical': canonical}
        return {'slug': run_id, 'stamp_utc': None, 'legacy': True, 'canonical': False}

    def run_dir(self, kind, team, slug, run_id, standalone):
        if kind not in ('team', 'playbook') or not isinstance(standalone, bool):
            raise ValueError('invalid kind or standalone (expected team/playbook and boolean)')
        for label, value in [('team', team), ('slug', slug), ('run_id', run_id)]:
            if not isinstance(value, str) or not self.segment_rx.fullmatch(value) or value in ('.', '..'):
                raise ValueError(f'invalid {label}')
        template = self.dirs['standalone' if standalone else 'playbook' if kind == 'playbook' else 'campus']
        stamp = self.parse_run_id(run_id)['stamp_utc']
        if '{YYYY-MM}' in template and stamp is None:
            raise ValueError('run_id has no timestamp from which to resolve YYYY-MM')
        values = {'YYYY-MM': stamp[:7] if stamp else '', 'team': self.playbook_owner if kind == 'playbook' else team,
                  'slug': slug, 'run_id': run_id}
        return {'run_dir': template.format_map(values)}


@lru_cache(maxsize=8)
def _load(path):
    return Resolver(path)


def _default():
    return _load(str(contract_path()))


def resolve_context(name): return _default().resolve_context(name)
def resolve_department(value): return _default().resolve_department(value)
def normalize_status(value): return _default().normalize_status(value)
def normalize_type(value): return _default().normalize_type(value)
def entry_file(value): return _default().entry_file(value)
def ledger_path(info): return _default().ledger_path(info)
def parse_run_id(value): return _default().parse_run_id(value)
def run_dir(kind, team, slug, run_id, standalone): return _default().run_dir(kind, team, slug, run_id, standalone)

FUNCTIONS = {'resolve_context': 1, 'resolve_department': 1, 'normalize_status': 1,
             'normalize_type': 1, 'entry_file': 1, 'ledger_path': 1, 'parse_run_id': 1, 'run_dir': 5}


def main(argv=None):
    try:
        argv = list(sys.argv[1:] if argv is None else argv)
        positional, override, pretty = [], None, False
        while argv:
            arg = argv.pop(0)
            if arg == '--':
                positional.extend(argv)
                break
            if arg == '--json':
                pretty = True
            elif arg == '--contract':
                if not argv:
                    raise ValueError('--contract needs a path')
                override = argv.pop(0)
            elif arg.startswith('--'):
                raise ValueError(f'unknown option: {arg}')
            else:
                positional.append(arg)
        if not positional or positional[0] not in FUNCTIONS:
            raise ValueError('unknown or missing function')
        function, *args = positional
        if len(args) != FUNCTIONS[function]:
            raise ValueError(f'{function} needs {FUNCTIONS[function]} argument(s)')
        if function == 'ledger_path':
            try: args[0] = json.loads(args[0])
            except ValueError: pass  # malformed seat content is an invalid seat, not CLI failure
        if function == 'run_dir':
            if args[-1] not in ('true', 'false'):
                raise ValueError('standalone must be true or false')
            args[-1] = args[-1] == 'true'
        result = getattr(Resolver(override), function)(*args)
        print(json.dumps(result, indent=2) if pretty else json.dumps(result))
        return 0
    except ContractError as exc:
        print(str(exc), file=sys.stderr)
        return 2
    except (ValueError, TypeError) as exc:
        print(f'USAGE: {exc}', file=sys.stderr)
        return 1


if __name__ == '__main__':
    raise SystemExit(main())

File: tools/contract.mjs
#!/usr/bin/env node
/** Read-only campus contract resolvers. Node >=18; no identity authority.
 * platform: codex; seat: codex
 */
import fs from 'node:fs';
import path from 'node:path';
import {fileURLToPath} from 'node:url';

export class ContractError extends Error {}
const ownFile = fileURLToPath(import.meta.url);
function contractPath(override) {
  return path.resolve(override || process.env.CAMPUS_CONTRACT ||
    path.join(path.dirname(ownFile), '..', 'contracts', 'campus.json'));
}
const object = value => value !== null && typeof value === 'object' && !Array.isArray(value);
const fullmatch = (rx, value) => {
  const m = typeof value === 'string' ? rx.exec(value) : null;
  return m && m[0] === value ? m : null;
};
// The generated patterns use Python named groups, Unicode decimal digits and
// dot-without-newline. Translate that subset once on load, never per parse.
function pythonRegex(source) {
  return new RegExp(source.replaceAll('(?P<', '(?<').replaceAll('\\d', '\\p{Decimal_Number}')
    .replaceAll('.+', '[^\\n]+'), 'u');
}
function requireShape(value) { if (!value) throw new Error('shape'); }

// Match Python strptime's %Y%m%d%H%M%S%f conversion, including its variable-width
// fields. Fixed slicing would disagree with the existing minter on old IDs.
const stampFields = pythonRegex('^(?<Y>\\d{4})(?<m>1[0-2]|0[1-9]|[1-9])' +
  '(?<d>3[0-1]|[1-2]\\d|0[1-9]|[1-9]| [1-9])(?<H>2[0-3]|[0-1]\\d|\\d| \\d)' +
  '(?<M>[0-5]\\d|\\d)(?<S>6[0-1]|[0-5]\\d|\\d)(?<f>[0-9]{1,6})$');
const decimal = /\p{Decimal_Number}/u;
function decimalNumber(value) {
  return Number([...value].map(char => {
    if (!decimal.test(char)) return char;
    const cp = char.codePointAt(0);
    let start = cp;
    while (start > 0 && decimal.test(String.fromCodePoint(start - 1))) start--;
    return String((cp - start) % 10);
  }).join(''));
}
function stamp(date, time, millis = '000') {
  const m = fullmatch(stampFields, date.replaceAll('-', '') +
    ([...time].length === 4 ? time + '00' : time) + millis);
  if (!m) return null;
  const g = m.groups;
  const [year, month, day, hour, minute, second] = ['Y', 'm', 'd', 'H', 'M', 'S'].map(k => decimalNumber(g[k]));
  if (year < 1 || year > 9999 || second > 59) return null;
  const value = new Date(0);
  value.setUTCFullYear(year, month - 1, day);
  value.setUTCHours(hour, minute, second, Number(g.f.padEnd(6, '0').slice(0, 3)));
  if (value.getUTCFullYear() !== year || value.getUTCMonth() !== month - 1 || value.getUTCDate() !== day) return null;
  return value.toISOString();
}

export class Resolver {
  constructor(override) {
    const filename = contractPath(override);
    try { this.data = JSON.parse(new TextDecoder('utf-8', {fatal: true, ignoreBOM: true}).decode(fs.readFileSync(filename))); }
    catch (e) { throw new ContractError(`${e.code === 'ENOENT' ? 'CONTRACT_MISSING' : 'CONTRACT_UNREADABLE'}: ${filename}`); }
    const c = this.data;
    if (!object(c) || c.contract_version !== '1.0')
      throw new ContractError('CONTRACT_VERSION_UNSUPPORTED: expected contract_version 1.0');
    try {
      this.files = c.context_aliases.files;
      requireShape(Array.isArray(this.files));
      for (const row of this.files) {
        requireShape(['concept', 'canonical', 'founding_writes'].every(k => typeof row[k] === 'string'));
        requireShape(Array.isArray(row.aliases) && row.aliases.every(a => typeof a === 'string'));
      }
      requireShape(JSON.stringify(c.context_aliases.resolution) === JSON.stringify([
        'canonical', 'legacy alias (in listed order)', 'case-insensitive match']));
      this.hosts = c.hosts;
      requireShape(object(this.hosts.clients) && object(this.hosts.entry_file_by_manifest_dir) &&
        Array.isArray(this.hosts.entry_file_read_names));
      for (const client of Object.values(this.hosts.clients))
        requireShape(typeof client.manifest_dir === 'string' && typeof client.entry_file === 'string');
      requireShape(Object.values(this.hosts.entry_file_by_manifest_dir).every(v => typeof v === 'string'));
      this.enums = c.registry_enums.enums;
      for (const key of ['department', 'status', 'type'])
        requireShape(Array.isArray(this.enums[key]) && this.enums[key].every(v => typeof v === 'string'));
      this.aliases = c.registry_enums.legacy_read_aliases.department;
      requireShape(object(this.aliases));
      this.seat = c.seat;
      this.seatRx = pythonRegex(this.seat.seat_id_pattern);
      for (const key of ['primary', 'satellite', 'unknown']) requireShape(typeof this.seat.ledger[key] === 'string');
      requireShape(Array.isArray(this.seat.roles) && typeof this.seat.lane.satellite === 'string');
      this.patterns = Object.entries(c.run_id.patterns).map(([name, source]) => [name, pythonRegex(source)]);
      requireShape(this.patterns.length && this.patterns[0][0] === 'canonical');
      this.segmentRx = pythonRegex(c.run_id.segment_pattern);
      this.dirs = c.run_id.run_dir;
      const pieces = this.dirs.campus.split('{team}');
      requireShape(pieces.length === 2);
      const [prefix, suffix] = pieces;
      requireShape(suffix && this.dirs.playbook.startsWith(prefix) && this.dirs.playbook.endsWith(suffix));
      this.playbookOwner = this.dirs.playbook.slice(prefix.length, -suffix.length);
      requireShape(fullmatch(this.segmentRx, this.playbookOwner) && typeof this.dirs.standalone === 'string');
    } catch (e) { throw new ContractError('CONTRACT_INVALID: required resolver data or regex invalid'); }
  }

  resolve_context(name) {
    let found = null, matched = null;
    for (const mode of ['canonical', 'alias', 'case-insensitive', 'concept']) {
      for (const row of this.files) {
        const names = [row.canonical, ...row.aliases];
        const concepts = [row.concept, ...row.concept.split('/'), ...names.map(n => n.replace(/\.[^.]*$/, ''))];
        const match = mode === 'canonical' ? name === row.canonical : mode === 'alias' ? row.aliases.includes(name) :
          mode === 'case-insensitive' ? names.some(n => n.toLowerCase() === name.toLowerCase()) :
          concepts.some(n => n.trim().toLowerCase() === name.toLowerCase());
        if (match) { found = row; matched = mode; break; }
      }
      if (found) break;
    }
    return {concept: found ? found.concept : null, canonical: found ? found.canonical : null,
      matched_by: matched, founding_writes: found ? found.founding_writes : null,
      read_order: found ? [found.canonical, ...found.aliases] : []};
  }
  resolve_department(value) {
    const canonical = this.enums.department.includes(value) ? value :
      Object.hasOwn(this.aliases, value) ? this.aliases[value] : null;
    const known = this.enums.department.includes(canonical);
    return {value, canonical: known ? canonical : null, legacy_alias: known && value !== canonical, known};
  }
  _normalize(value, kind) {
    if (this.enums[kind].includes(value)) return {value, canonical: value, known: true};
    for (const canonical of this.enums[kind])
      if (value.toLowerCase() === canonical.toLowerCase())
        return {value, canonical, known: true, matched_by: 'case-insensitive'};
    return {value, canonical: null, known: false};
  }
  normalize_status(value) { return this._normalize(value, 'status'); }
  normalize_type(value) { return this._normalize(value, 'type'); }
  entry_file(value) {
    const client = Object.hasOwn(this.hosts.clients, value) ? this.hosts.clients[value] : null;
    const directory = client ? client.manifest_dir : value;
    const entry = Object.hasOwn(this.hosts.entry_file_by_manifest_dir, directory) ? this.hosts.entry_file_by_manifest_dir[directory] : null;
    return {manifest_dir: entry !== null ? directory : null, entry_file: entry, read_names: [...this.hosts.entry_file_read_names]};
  }
  ledger_path(info) {
    const valid = Boolean(object(info) && typeof info.seat === 'string' && fullmatch(this.seatRx, info.seat) && this.seat.roles.includes(info.role));
    const state = valid ? 'valid' : info === null ? 'missing' : 'invalid';
    const role = valid ? info.role : 'satellite', seat = valid ? info.seat : 'unknown';
    const ledger = this.seat.ledger[valid ? role : 'unknown'].replaceAll('{seat}', seat);
    const lane = role === 'satellite' ? this.seat.lane.satellite.replaceAll('{seat}', seat) : null;
    return {state, role, seat, ledger, lane};
  }
  parse_run_id(runId) {
    for (const [name, rx] of this.patterns) {
      const match = fullmatch(rx, runId);
      if (!match) continue;
      const v = match.groups, timestamp = stamp(v.date, v.time, v.millis || '000');
      if (timestamp === null) break;
      const canonical = name === 'canonical';
      return {slug: v.slug || runId, stamp_utc: timestamp, legacy: !canonical, canonical};
    }
    return {slug: runId, stamp_utc: null, legacy: true, canonical: false};
  }
  run_dir(kind, team, slug, runId, standalone) {
    if (!['team', 'playbook'].includes(kind) || typeof standalone !== 'boolean')
      throw new Error('invalid kind or standalone (expected team/playbook and boolean)');
    for (const [label, value] of [['team', team], ['slug', slug], ['run_id', runId]])
      if (typeof value !== 'string' || !fullmatch(this.segmentRx, value) || ['.', '..'].includes(value)) throw new Error(`invalid ${label}`);
    const template = this.dirs[standalone ? 'standalone' : kind === 'playbook' ? 'playbook' : 'campus'];
    const timestamp = this.parse_run_id(runId).stamp_utc;
    if (template.includes('{YYYY-MM}') && timestamp === null) throw new Error('run_id has no timestamp from which to resolve YYYY-MM');
    const values = {'YYYY-MM': timestamp ? timestamp.slice(0, 7) : '',
      team: kind === 'playbook' ? this.playbookOwner : team, slug, run_id: runId};
    return {run_dir: template.replace(/\{([^}]+)\}/g, (_, key) => {
      if (!Object.hasOwn(values, key)) throw new Error(`unknown template field ${key}`);
      return values[key];
    })};
  }
}

const cache = new Map();
function defaultResolver() {
  const filename = contractPath();
  if (!cache.has(filename)) cache.set(filename, new Resolver(filename));
  return cache.get(filename);
}
export const resolve_context = name => defaultResolver().resolve_context(name);
export const resolve_department = value => defaultResolver().resolve_department(value);
export const normalize_status = value => defaultResolver().normalize_status(value);
export const normalize_type = value => defaultResolver().normalize_type(value);
export const entry_file = value => defaultResolver().entry_file(value);
export const ledger_path = value => defaultResolver().ledger_path(value);
export const parse_run_id = value => defaultResolver().parse_run_id(value);
export const run_dir = (...args) => defaultResolver().run_dir(...args);
export const FUNCTIONS = {resolve_context: 1, resolve_department: 1, normalize_status: 1,
  normalize_type: 1, entry_file: 1, ledger_path: 1, parse_run_id: 1, run_dir: 5};

function ascii(text) { return text.replace(/[\u007f-\uffff]/g, c => '\\u' + c.charCodeAt(0).toString(16).padStart(4, '0')); }
function compact(value) {
  if (Array.isArray(value)) return '[' + value.map(compact).join(', ') + ']';
  if (object(value)) return '{' + Object.entries(value).map(([k, v]) => JSON.stringify(k) + ': ' + compact(v)).join(', ') + '}';
  return JSON.stringify(value);
}
export function main(argv = process.argv.slice(2)) {
  try {
    argv = [...argv];
    const positional = [];
    let override = null, pretty = false;
    while (argv.length) {
      const arg = argv.shift();
      if (arg === '--') { positional.push(...argv); break; }
      if (arg === '--json') pretty = true;
      else if (arg === '--contract') {
        if (!argv.length) throw new Error('--contract needs a path');
        override = argv.shift();
      } else if (arg.startsWith('--')) throw new Error(`unknown option: ${arg}`);
      else positional.push(arg);
    }
    const fn = positional.shift();
    if (!Object.hasOwn(FUNCTIONS, fn)) throw new Error('unknown or missing function');
    if (positional.length !== FUNCTIONS[fn]) throw new Error(`${fn} needs ${FUNCTIONS[fn]} argument(s)`);
    if (fn === 'ledger_path') { try { positional[0] = JSON.parse(positional[0]); } catch {} }
    if (fn === 'run_dir') {
      if (!['true', 'false'].includes(positional.at(-1))) throw new Error('standalone must be true or false');
      positional[positional.length - 1] = positional.at(-1) === 'true';
    }
    const result = new Resolver(override)[fn](...positional);
    process.stdout.write(ascii(pretty ? JSON.stringify(result, null, 2) : compact(result)) + '\n');
    return 0;
  } catch (e) {
    process.stderr.write((e instanceof ContractError ? e.message : `USAGE: ${e.message}`) + '\n');
    return e instanceof ContractError ? 2 : 1;
  }
}
if (process.argv[1] && path.resolve(process.argv[1]) === ownFile) process.exitCode = main();
